text/0000-offset_of.md

```diff
@@ -0,0 +1,480 @@
+    metadata: [u64; 3],
+    datadata: [u8; 1024 * 1024 * 32],
+}
+const METADATA_OFFSET: usize = offset_of!(Buffer, metadata);
+```
+
+## Limitations
+
+There are a few limitations worth mentioning. Some of these may be relaxed in
+the future, however.
+
+1. Perhaps unsurprisingly, it obeys privacy, so both the type and field you call
+   `offset_of!` on must be visible to the code calling `offset_of!`.
+
+2. The type holding the field must be `Sized`, so trying to compute where the
+   slice begins in something like `offset_of!((i32, [u32]), 1)` isn't supported.
+
+3. Compared to `offsetof` in C and C++, you can't access nested fields/arrays.
+   That is, instead of `offset_of!(Foo, quank.zoop.2.quank[4])`, you'll have to
+   compute the offsets of each step manually, and sum them.
+
+4. Finally, types other than tuples, structs, and unions are not currently
+   unsupported.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`offset_of` is a new macro exported from `core::mem` which has an signature
+similar to the following:
+
+```rs
+pub macro offset_of($Container:ty, $field:tt $(,)?) {
+    // ...implementation defined...
+}
+```
+
+Invoking this macro expands to a constant expression of type `usize`, which
+evaluates to the offset in bytes from the beginning of `$Container` where
+`$field` is found.
+
+`$Container` must be visible and must be or resolve to one of the following
+types:
+
+1. A `struct` or `union` type with either named or anonymous/tuple-style fields.
+
+    In this case, `$field` must share a name or tuple index with a field which:
+    - Exists on `$Container`.
+    - Is visible at the location where `offset_of!` is invoked (but there is no
+      requirement that fields other than than `$field` be visible there)
+
+2. An anonymous tuple type.
+
+    In this case, `$field` must be a tuple index (that is, an integer literal)
+    that exists on the tuple type in question.
+
+Use on other types is an error, although this may be relaxed in some cases in
+the future (see the [Future possibilities][future-possibilities] section).
+
+As a note: the implementation is strongly encouraged to not have runtime
+resource usage dependent on the values of `$Container` or `$field`. In
+particular, the implementation should not allocate space for an instance of
+`$Container` on the runtime stack.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. This exposes layout information at compile time which is otherwise not
+   exposed until runtime. This can cause compatibility hazards similar to
+   `mem::size_of` or `mem::align_of`, but plausibly greater as it provides even
+   more information.
+
+    That said, this API allows querying information which (if needed at compile
+    time) would otherwise be hard-coded, so in some cases it may reduce the risk
+    of a compatibility hazard.
+
+2. This is a low level feature that most code won't need to use, so perhaps it
+   is better off left out.
```

Comment:
I don't think that "is low level" should be put down as a general justification for Rust to not to do something, even if by accepting this RFC we're rejecting the concern for this particular case.

Reply:
Yeah I agree. That said, a concern that most code won't use it is probably valid.

I think in this case it's probably widely used enough -- at one point I *think* I remember @saethlin mentioning that "incorrect `offset_of` impl" among the most common (possibly just the most common outright) causes of UB found when running miri on the top 10k crates from crates.io.

(Of course, that's a bit of a biased set, but hearing that convinced me to finish this up -- although it took a while to get to it).

Reply:
I've reworded this to be closer to what I meant. That said, this is really a pseudo-straw-man based on a half-remembered argument. Maybe I should remove the point.